lte/gateway/c/session_manager/LocalEnforcer.cpp

```diff
@@ -558,84 +558,99 @@ void LocalEnforcer::populate_final_action_install_info(
 void LocalEnforcer::start_final_unit_action_flows_install(
     SessionMap& session_map, const FinalActionInstallInfo final_action_info,
     SessionUpdate& session_update) {
+  const auto &imsi       = final_action_info.imsi,
+             &session_id = final_action_info.session_id;
+  // First check if the UE IPv4 field is filled out & ready to use
+  SessionSearchCriteria criteria(imsi, IMSI_AND_SESSION_ID, session_id);
+  auto session_it = session_store_.find_session(session_map, criteria);
+  if (!session_it) {
+    MLOG(MERROR) << session_id
+                 << " not found when trying to install final unit action flows";
+    return;
+  }
+  auto ip = (**session_it)->get_config().common_context.ue_ipv4();
+  if (!ip.empty()) {
+    complete_final_unit_action_flows_install(
+        session_map, ip, final_action_info, session_update);
+    return;
+  }
+
+  // If UE IPv4 does not exist in the context, fetch it from DirectoryD
   MLOG(MDEBUG) << "Fetching Subscriber IP address from DirectoryD for "
-               << final_action_info.session_id;
+               << session_id;
   directoryd_client_->get_directoryd_ip_field(
-      final_action_info.imsi,
-      [this, final_action_info](Status status, DirectoryField resp) {
+      imsi, [this, final_action_info](Status status, DirectoryField resp) {
         // This call back gets executed in the DirectoryD client thread, but
         // we want to run the session update logic in the main thread.
-        evb_->runInEventBaseThread([this, final_action_info, status, resp]() {
-          complete_final_unit_action_flows_install(
-              status, resp, final_action_info);
-        });
+        if (!status.ok()) {
+          MLOG(MERROR) << "Could not fetch IP info for "
+                       << final_action_info.session_id
+                       << ". Failing final action flow install error: "
+                       << status.error_message();
+          return;
+        }
+        evb_->runInEventBaseThread(
```

Comment:
so for the cwf case, the install of the flow will not happen immediately, but on the next available slot on the event thread. 

So nn this meantime, what is the state of the session?

Reply:
Yes. The state will be that the credit has run out and the SERVICE_STATE will indicate -> REDIRECTED. The redirect rule will only get added to session state only after we reach this next available slot.